File: app/src/main/java/ch/olivsoft/android/blindman/Obstacle.java
package ch.olivsoft.android.blindman;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.PaintDrawable;
import android.graphics.drawable.shapes.Shape;

public class Obstacle extends PaintDrawable {

    private boolean hit = false;
    private boolean hidden = false;

    Obstacle(int ix, int iy, int size) {
        super();
        // The allocated shape in a PaintDrawable is
        // a rectangle defined by its bounds
        setBounds(ix * size, iy * size, (ix + 1) * size, (iy + 1) * size);
        setCornerRadius(0.1f * size);
        getPaint().setAntiAlias(true);
    }

    // Some get and set.
    // We do NOT use the inherited setVisible method and hope
    // that no-one ever calls isVisible. That is also why we use
    // "hidden" instead of "visible" as property of an obstacle.
    // We implement many simple methods in order to relieve
    // the caller from taking care of the obstacle's state logic.
    boolean isHit() {
        return hit;
    }

    void setHit() {
        hit = true;
        hidden = false;
    }

    void setHidden() {
        hidden = true;
    }

    void setVisibleIfHit() {
        hidden = !hit;
    }

    void setVisible() {
        hidden = false;
    }

    // Convenience method for intersection check
    boolean intersects(Rect r) {
        return Rect.intersects(getBounds(), r);
    }

    // Draw only visible obstacles, and in the right color
    @Override
    protected void onDraw(Shape shape, Canvas canvas, Paint paint) {
        if (hidden)
            return;
        paint.setColor(ColoredPart.OBSTACLE.color);
        super.onDraw(shape, canvas, paint);
    }
}

File: app/src/main/java/ch/olivsoft/android/blindman/Obstacle.kt
package ch.olivsoft.android.blindman

import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.PaintDrawable
import androidx.compose.ui.graphics.toComposeIntRect
import androidx.compose.ui.unit.IntRect

class Obstacle(ix: Int, iy: Int, size: Int) : PaintDrawable() {
    // We implement all necessary methods in order to relieve
    // the caller from taking care of the obstacle's state logic.

    private var hit = false
    private var visible = true
    private val composeRect: IntRect

    init {
        // The allocated shape in a PaintDrawable is
        // a rectangle defined by its bounds
        setBounds(
            ix * size, iy * size,
            (ix + 1) * size, (iy + 1) * size
        )
        setCornerRadius(0.1f * size)
        paint.isAntiAlias = true
        composeRect = bounds.toComposeIntRect()
    }

    fun isHit(): Boolean {
        return hit
    }

    fun setHit() {
        hit = true
        visible = true
    }

    fun setHidden() {
        visible = false
    }

    fun setVisibleIfHit() {
        visible = hit
    }

    fun setVisible() {
        visible = true
    }

    // Convenience methods for intersection checks
    fun intersects(r: Rect): Boolean {
        return Rect.intersects(bounds, r)
    }

    fun overlaps(r: IntRect): Boolean {
        return composeRect.overlaps(r)
    }

    // Draw visible obstacles in the current color
    override fun draw(canvas: Canvas) {
        draw(canvas, ColoredPart.OBSTACLE.color)
    }

    fun draw(canvas: Canvas, color: Int) {
        if (visible) {
            paint.color = color
            super.draw(canvas)
        }
    }
}
